### PR TITLE
TUP-20309 : handle bulk processing in DI

### PR DESCRIPTION
**Pull request by akhabali**
### Requirements

* Any code change adding any logic **MUST** be tested through a unit test executed with the default build
* Any API addition **MUST** be done with a documentation update if relevant

### CODE REVIEW

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/visitor/ModelVisitor.java [hunk 1]
```diff
@@ -158,14 +157,11 @@ private void validateProcessor(final Class<?> input) {
...
 
-        Stream
-                .of(input.getMethods())
-                .filter(m -> m.isAnnotationPresent(BeforeGroup.class) || m.isAnnotationPresent(AfterGroup.class))
-                .forEach(m -> {
-                    if (m.getParameterCount() > 0) {
-                        throw new IllegalArgumentException(m + " must not have any parameter");
-                    }
-                });
+        Stream.of(input.getMethods()).filter(m -> m.isAnnotationPresent(BeforeGroup.class)).forEach(m -> {
```
Comment (rmannibucau) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/visitor/ModelVisitor.java]: please validate @AfterGroup only has @Output params if it has any param

File: component-runtime-testing/component-runtime-junit/src/main/java/org/talend/sdk/component/junit/BaseComponentsHandler.java [hunk 2]
```diff
@@ -65,6 +65,7 @@
 import org.talend.sdk.component.runtime.manager.ContainerComponentRegistry;
```
Comment (rmannibucau) [inline component-runtime-testing/component-runtime-junit/src/main/java/org/talend/sdk/component/junit/BaseComponentsHandler.java]: ensure this feature is tested please

File: component-runtime-di/src/main/java/org/talend/sdk/component/runtime/di/BaseIOHandler.java [hunk 3]
```diff
@@ -54,38 +53,61 @@ public void init(final Collection<String> branchesOrder) {
...
         return "__default__".equals(name) ? "FLOW" : name;
     }
 
-    @AllArgsConstructor
-    @Data
+    @RequiredArgsConstructor
     static class IO<T> {
 
-        private final AtomicReference<T> value;
+        private final Queue<T> values = new ConcurrentLinkedQueue<>();
```
Comment (rmannibucau) [inline component-runtime-di/src/main/java/org/talend/sdk/component/runtime/di/BaseIOHandler.java]: can be a linkedlist I think

File: component-runtime-di/src/main/java/org/talend/sdk/component/runtime/di/BaseIOHandler.java [hunk 4]
```diff
@@ -54,38 +53,61 @@ public void init(final Collection<String> branchesOrder) {
...
 
         private final Class<T> type;
 
-        private boolean mutated;
+        private final boolean mutated;
```
Comment (rmannibucau) [inline component-runtime-di/src/main/java/org/talend/sdk/component/runtime/di/BaseIOHandler.java]: if not needed please drop it

File: component-runtime-di/src/main/java/org/talend/sdk/component/runtime/di/BaseIOHandler.java [hunk 5]
```diff
@@ -18,14 +18,14 @@
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.Iterator;
+import java.util.LinkedList;
 import java.util.Map;
+import java.util.Queue;
 import java.util.TreeMap;
-import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.ConcurrentLinkedQueue;
```
Comment (rmannibucau) [inline component-runtime-di/src/main/java/org/talend/sdk/component/runtime/di/BaseIOHandler.java]: needed?

Comment (akhabali) [reply component-runtime-di/src/main/java/org/talend/sdk/component/runtime/di/BaseIOHandler.java]: absolutely not. will remove

**Pull request merged**